src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,97 @@
+
+     /**
+    * Verifies that the Google token is legitimate. Then creates a 
+    * cookie storing the Session ID and stores the login information in the user entity in datastore.
+    *
+    * User Entity:
+    *    username      The user's full name (from their google account)
+    *    userID        A unique string that can identify each user (from their google account)
+    *    SessionID     A unique ID that changes every time a user logs in
+    *    quiz_timing   A value that tracks the last time the user took a quiz. (Used for game)
+    *    blobkey       A value that can be used to access a picture uploaded by the user
+    *    currentGame   A string representing the game the user is currently in
+    *    score         The score of the user (int)
+    *
+    *
+    * If a User Entity already exists in the database, the only part of the entry that is updated
+    * is the SessionID property. 
+    *
+    *<p>
+    *@param  idToken  a token used to identify the user.
+    *@return          a cookie that contains the session ID
+    *@return          (ONLY IF ONE DOESNT ALREADY EXIST) a user entity in datastore
+    */
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+     
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    JsonFactory jsonFactory = new JacksonFactory();
+    HttpTransport transport = new NetHttpTransport(); 
+
+    GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
+            .setAudience(Collections.singletonList("610251294652-9ojdhjhh8kpcdvdbmrvp9nkevgr2n2d8.apps.googleusercontent.com"))
+            .build();
+
+    Enumeration<String> headers = request.getHeaders("idtoken");
+    String idTokenString = null;
+    while(headers.hasMoreElements()){
+        idTokenString = headers.nextElement();
+         //header passed in should always have ONLY 1 element (just the id token)
+         //therefore, id token should always be the last element.
+    }
+
+    try{
+        GoogleIdToken idToken = verifier.verify(idTokenString);
+        if (idToken != null) {
+          //create coookie w session ID
+          String sessionID = request.getSession(true).getId(); 
+          Cookie sessionIDcookie = CookieUtils.createSessionIDCookie(request.getSession(false), sessionID);
+          response.addCookie(sessionIDcookie);
+
+          Payload payload = idToken.getPayload();
+          //get user info for user entity
+          String userId = payload.getSubject();
+          String username = (String) payload.get("name");
+
+          //send user data to datastore
+          Entity userEntity = UserUtils.getEntityFromDatastore("user", "userID", userId, datastore);
+          if(userEntity == null){
+            datastore.put(UserUtils.initializeUser(userId, username, sessionID));
```

Comment:
Do we need to check if this function returns null? Or is this already in a test for UserUtils

Reply:
yea! i refactored it a little so now the initialize user fxn just creates the user. you now have to check if the user exists before putting it into the database.